pkg/compact/retention.go

```diff
@@ -40,3 +41,35 @@ func ApplyDefaultRetentionPolicy(ctx context.Context, logger log.Logger, bkt obj
 	level.Info(logger).Log("msg", "default retention apply done")
 	return nil
 }
+
+// Apply removes blocks of resolution older than rententionDuration based on blocks MaxTime.
+func ApplyRetentionPolicyByResolution(ctx context.Context, logger log.Logger, bkt objstore.Bucket, retentionDuration time.Duration, resolution int64) error {
+	level.Info(logger).Log("msg", fmt.Sprintf("start retention of resolution %d", resolution))
+	if err := bkt.Iter(ctx, "", func(name string) error {
+		id, ok := block.IsBlockDir(name)
+		if !ok {
+			return nil
+		}
+		m, err := block.DownloadMeta(ctx, logger, bkt, id)
+		if err != nil {
+			return errors.Wrap(err, "download metadata")
+		}
+
+		if m.Thanos.Downsample.Resolution == resolution {
```

Comment:
TBH as this is a little bit simpler, we could avoid iterating 3 times and we could do all here just once. (we have retentionByRes map). WDYT?

Reply:
Yes it saves time and reduces load to the s3.
If we avoid iterating multiple times the metadata does not need to get downloaded multiple times.